File: examples/kotlin/grammar/stock-record/kotlin/cott_impl/curriculum/stock_record/value_stock_record.kt
package cott_impl.curriculum.stock_record

internal fun value_stock_record(record: curriculum.stock_record.StockRecord): cott_runtime.CottResult<kotlin.Double, curriculum.stock_record.StockRecordError> {
    if (record.name.isEmpty()) {
        return cott_runtime.Err(curriculum.stock_record.StockRecordError.EmptyName)
    }
    if (record.shares < 0L) {
        return cott_runtime.Err(curriculum.stock_record.StockRecordError.NegativeShares)
    }
    if (!(-curriculum.stock_record.MAX_F64 <= record.price && record.price <= curriculum.stock_record.MAX_F64)) {
        return cott_runtime.Err(curriculum.stock_record.StockRecordError.NonFinitePrice)
    }
    if (record.price < 0.0) {
        return cott_runtime.Err(curriculum.stock_record.StockRecordError.NegativePrice)
    }
    return curriculum.stock_record.value_record(record)
}

File: examples/grammar/stock-record/python/_cott_impl/curriculum/stock_record/value_stock_record.py
from cott_runtime import Err, F64, Result
from curriculum.stock_record import value_record
from curriculum.stock_record_types import MAX_F64, StockRecord, StockRecordError, StockRecordError_EmptyName, StockRecordError_NegativePrice, StockRecordError_NegativeShares, StockRecordError_NonFinitePrice


def value_stock_record(record: StockRecord) -> Result[F64, StockRecordError]:
    if len(record.name) == 0:
        return Err(error=StockRecordError_EmptyName())
    if record.shares < 0:
        return Err(error=StockRecordError_NegativeShares())
    if not (-MAX_F64 <= record.price <= MAX_F64):
        return Err(error=StockRecordError_NonFinitePrice())
    if record.price < 0:
        return Err(error=StockRecordError_NegativePrice())
    return value_record(record)
